### PR TITLE
AuthMe improvements and fixes

### PR DESCRIPTION
**Pull request by tadhgboyle**
This PR brings some changes to the AuthMe system:
- Add check authme is setup properly before allowing users to enter username/password
- Add support for making users enter profile fields when they enter their email on the second step of authme registration
- Use `DB` class instead of `mysqli` for connecting to authme db
- Allow users to configure if they get their authme password synced upon login or not

### CODE REVIEW

File: custom/templates/DefaultRevamp/authme_email.tpl [hunk 1]
```diff
@@ -31,17 +31,64 @@
                     </div>
                 </div>
                 <form class="ui form" action="" method="post" id="form-authme-email">
-                    <div class="field">
-                        <label for="inputEmail">{$EMAIL}</label>
-                        <input type="email" id="inputEmail" name="email" placeholder="{$EMAIL}">
-                    </div>
+                    {assign var=counter value=1}
```
Comment (tadhgboyle) [inline custom/templates/DefaultRevamp/authme_email.tpl:34]: We should look into reusable tpl files 🤔 😅

**Review by partydragen — COMMENTED** [review]
I believe this caused duplication issue, As module column NULL being always as unique same, it dont care if key already exist when module is null

File: custom/templates/DefaultRevamp/authme_email.tpl (hunk 1 above)
Comment (partydragen) [reply custom/templates/DefaultRevamp/authme_email.tpl:34]: make the fields stuff into a include tpl? Yea i thinked about doing that when i made it, same as few of my modules will also use this

**Comment by partydragen** [comment]
> I believe this caused duplication issue, As module column NULL being always as unique same, it dont care if key already exist when module is null

File: modules/Core/pages/authme_connector.php [hunk 2]
```diff
@@ -28,139 +28,153 @@
...
-                    ],
-                    'email' => [
-                        Validate::REQUIRED => true,
-                        Validate::MIN => 4,
-                        Validate::MAX => 64,
-                        Validate::UNIQUE => 'users'
-                    ]
-                ]);
-            } else {
-                $validation = Validate::check($_POST, [
```
Comment (partydragen) [inline modules/Core/pages/authme_connector.php:46]: Hm we still need username validation check, I see it never was here but we still need to make sure username is unique, same rules as nickname

File: modules/Core/pages/authme_connector.php [hunk 3]
```diff
@@ -28,139 +28,153 @@
...
-                ]);
+                        Validate::MAX => (is_null($field->length) ? 1024 : $field->length)
+                    ];
+                }
             }
 
+            $validation = Validate::check($_POST, $to_validation);
+
             $validation->messages([
-                'username' => [
```
Comment (partydragen) [inline modules/Core/pages/authme_connector.php:57]: Then you need to keep this aswell

File: modules/Core/pages/authme_connector.php (hunk 2 above)
Comment (tadhgboyle) [reply modules/Core/pages/authme_connector.php:46]: Great catch

File: custom/panel_templates/Default/integrations/minecraft/minecraft_authme.tpl [hunk 4]
```diff
@@ -65,54 +65,49 @@
...
-                                    <label for="inputAuthmeSync" class="custom-control-label">
-                                        {$AUTHME_PASSWORD_SYNC} <span class="badge badge-info"><i class="fa fa-question-circle" data-container="body" data-toggle="popover" data-placement="top" title="{$INFO}" data-content="{$AUTHME_PASSWORD_SYNC_HELP}"></i></span>
-                                    </label>
+                                    <input type="text" class="form-control" name="db_table" value="{if isset($AUTHME_DB_DETAILS['table'])}{$AUTHME_DB_DETAILS['table']|escape}{else}authme{/if}">
                                 </div>
                                 <div class="form-group">
                                     <input type="hidden" name="token" value="{$TOKEN}">
                                     <input type="submit" value="{$SUBMIT}" class="btn btn-primary">
+                                    <button class="btn btn-info" id="testConnection">Test connection</button>
```
Comment (partydragen) [inline custom/panel_templates/Default/integrations/minecraft/minecraft_authme.tpl]: Missing translation

File: modules/Core/pages/authme_connector.php [hunk 5]
```diff
@@ -26,141 +27,179 @@
...
+                        'nickname' => $nickname,
+                        'password' => $_SESSION['authme']['pass'],
+                        'pass_method' => $_SESSION['authme']['hash'],
+                        'joined' => date('U'),
+                        'email' => $email,
+                        'reset_code' => $code,
+                        'lastip' => $ip,
+                        'last_online' => date('U'),
+                        'language_id' => $language_id,
+                        'authme_registered' => true,
```
Comment (partydragen) [inline modules/Core/pages/authme_connector.php]: switch to register_method instead or something then we could also use it for oauth_{service} etc

File: modules/Core/pages/authme_connector.php [hunk 6]
```diff
@@ -26,141 +27,179 @@
...
+                        'authme_sync_password' => Input::get('authme_sync_password') === 'on',
+                    ]);
+
```
Comment (partydragen) [inline modules/Core/pages/authme_connector.php:153]: Missing to fire new user event

**Review by partydragen — CHANGES_REQUESTED** [review]

**Review by partydragen — APPROVED** [review]

**Pull request merged**